File: backend/app/services/order_service.py
from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import Session, selectinload

from app.core.exceptions import ConflictError, InsufficientStockError, NotFoundError
from app.models.order import Order, OrderStatus
from app.models.order_item import OrderItem
from app.models.order_status_history import OrderStatusHistory
from app.models.product_variant import ProductVariant
from app.models.user import User
from app.schemas.order import OrderCreate, OrderStatusUpdate
from app.utils.order_number_generator import generate_order_number
from app.utils.pagination import PageParams, build_meta

_MAX_ORDER_NUMBER_RETRIES = 3


def _load(query):
    return query.options(
        selectinload(Order.items).selectinload(OrderItem.product_variant).selectinload(
            ProductVariant.product
        )
    )


def create_order(db: Session, payload: OrderCreate) -> Order:
    """Décrémente le stock et crée la commande dans une seule transaction,
    avec verrouillage de ligne pour éviter la survente en cas de commandes
    simultanées sur la même variante.
    """
    for attempt in range(_MAX_ORDER_NUMBER_RETRIES):
        try:
            with db.begin_nested():
                total_amount = 0
                order_items: list[OrderItem] = []

                for item_in in payload.items:
                    variant = (
                        db.query(ProductVariant)
                        .filter(ProductVariant.id == item_in.product_variant_id)
                        .with_for_update()
                        .first()
                    )
                    if variant is None:
                        raise NotFoundError(
                            f"Variante produit {item_in.product_variant_id} introuvable."
                        )
                    if variant.stock_quantity < item_in.quantity:
                        raise InsufficientStockError(
                            f"Stock insuffisant pour {variant.product.name} "
                            f"({variant.color}, {variant.size})."
                        )

                    variant.stock_quantity -= item_in.quantity
                    unit_price = variant.product.base_price
                    subtotal = unit_price * item_in.quantity
                    total_amount += subtotal

                    order_items.append(
                        OrderItem(
                            product_variant_id=variant.id,
                            quantity=item_in.quantity,
                            unit_price=unit_price,
                            subtotal=subtotal,
                        )
                    )

                order = Order(
                    order_number=generate_order_number(db),
                    customer_name=payload.customer_name,
                    customer_phone=payload.customer_phone,
                    customer_address=payload.customer_address,
                    status=OrderStatus.pending,
                    payment_method=payload.payment_method,
                    total_amount=total_amount,
                    items=order_items,
                )
                db.add(order)
                db.flush()

                db.add(
                    OrderStatusHistory(
                        order_id=order.id, old_status=None, new_status=OrderStatus.pending
                    )
                )

            db.commit()
            db.refresh(order)
            return order

        except IntegrityError:
            db.rollback()
            if attempt == _MAX_ORDER_NUMBER_RETRIES - 1:
                raise
            continue


def get_order(db: Session, order_id: int) -> Order:
    order = _load(db.query(Order)).filter(Order.id == order_id).first()
    if order is None:
        raise NotFoundError("Commande introuvable.")
    return order


def track_order(db: Session, *, order_number: str, phone: str) -> Order:
    order = (
        _load(db.query(Order))
        .filter(Order.order_number == order_number, Order.customer_phone == phone)
        .first()
    )
    if order is None:
        raise NotFoundError("Aucune commande ne correspond à ces informations.")
    return order


def list_orders(
    db: Session,
    *,
    status: OrderStatus | None,
    date_from,
    date_to,
    page_params: PageParams,
) -> tuple[list[Order], dict]:
    query = db.query(Order)
    if status:
        query = query.filter(Order.status == status)
    if date_from:
        query = query.filter(Order.created_at >= date_from)
    if date_to:
        query = query.filter(Order.created_at <= date_to)

    total = query.count()
    items = (
        query.order_by(Order.created_at.desc())
        .offset(page_params.offset)
        .limit(page_params.page_size)
        .all()
    )
    meta = build_meta(page=page_params.page, page_size=page_params.page_size, total=total)
    return items, meta


_ALLOWED_TRANSITIONS: dict[OrderStatus, set[OrderStatus]] = {
    OrderStatus.pending: {OrderStatus.processing, OrderStatus.cancelled},
    OrderStatus.processing: {OrderStatus.shipped, OrderStatus.cancelled},
    OrderStatus.shipped: {OrderStatus.delivered},
    OrderStatus.delivered: set(),
    OrderStatus.cancelled: set(),
}


def update_order_status(
    db: Session, order_id: int, payload: OrderStatusUpdate, *, changed_by: User
) -> Order:
    order = get_order(db, order_id)

    if payload.status != order.status and payload.status not in _ALLOWED_TRANSITIONS[order.status]:
        raise ConflictError(
            f"Transition de statut invalide : {order.status.value} → {payload.status.value}."
        )

    old_status = order.status
    order.status = payload.status
    db.add(
        OrderStatusHistory(
            order_id=order.id,
            old_status=old_status,
            new_status=payload.status,
            changed_by=changed_by.id,
        )
    )
    db.commit()
    db.refresh(order)
    return order

File: frontend/src/services/api/orderService.js
import { env } from "../../config/env";
import { mockOrders } from "../../mocks/orders";
import { paginate, simulateDelay } from "../../mocks";
import { loadMockState, saveMockState } from "../../mocks/persistence";
import httpClient from "./httpClient";
import { decrementMockVariantStock, findMockVariant } from "./productService";

let mockStore = loadMockState("orders", mockOrders);
let mockOrderSeq = loadMockState("order_seq", mockStore.length + 1);

function persist() {
  saveMockState("orders", mockStore);
  saveMockState("order_seq", mockOrderSeq);
}

function mockOrderError(status, code, message) {
  const error = new Error(message);
  error.response = { status, data: { error: { code, message } } };
  return error;
}

export async function createOrder(payload) {
  if (env.useMocks) {
    for (const item of payload.items) {
      const found = findMockVariant(item.product_variant_id);
      if (!found) {
        throw mockOrderError(404, "not_found", "Variante produit introuvable.");
      }
      if (found.variant.stock_quantity < item.quantity) {
        throw mockOrderError(
          409,
          "insufficient_stock",
          `Stock insuffisant pour ${found.product.name} (${found.variant.color}, ${found.variant.size}).`,
        );
      }
    }

    const items = payload.items.map((item, i) => {
      const { product, variant } = decrementMockVariantStock(item.product_variant_id, item.quantity);
      const unit_price = product.base_price;
      const subtotal = (parseFloat(unit_price) * item.quantity).toFixed(2);
      return {
        id: Date.now() + i,
        product_variant_id: item.product_variant_id,
        product_name: product.name,
        size: variant.size,
        color: variant.color,
        quantity: item.quantity,
        unit_price,
        subtotal,
      };
    });
    const now = new Date();
    const order = {
      id: Date.now(),
      order_number: `NG-${now.getFullYear()}-${String(mockOrderSeq++).padStart(6, "0")}`,
      customer_name: payload.customer_name,
      customer_phone: payload.customer_phone,
      customer_address: payload.customer_address,
      payment_method: payload.payment_method,
      status: "pending",
      items,
      total_amount: items.reduce((sum, i) => sum + parseFloat(i.subtotal), 0).toFixed(2),
      created_at: now.toISOString(),
    };
    mockStore = [order, ...mockStore];
    persist();
    return simulateDelay({ data: order });
  }
  const { data } = await httpClient.post("/orders", payload);
  return data;
}

export async function trackOrder(orderNumber, phone) {
  if (env.useMocks) {
    const order = mockStore.find(
      (o) => o.order_number === orderNumber && o.customer_phone === phone,
    );
    if (!order) {
      throw mockOrderError(404, "not_found", "Aucune commande trouvée.");
    }
    return simulateDelay({ data: order });
  }
  const { data } = await httpClient.get("/orders/track", { params: { order_number: orderNumber, phone } });
  return data;
}

export async function listOrders(params = {}) {
  if (env.useMocks) {
    let filtered = [...mockStore];
    if (params.status) filtered = filtered.filter((o) => o.status === params.status);
    filtered.sort((a, b) => new Date(b.created_at) - new Date(a.created_at));
    return simulateDelay(paginate(filtered, params.page || 1, params.page_size || 20));
  }
  const { data } = await httpClient.get("/orders", { params });
  return data;
}

export async function getOrder(id) {
  if (env.useMocks) {
    return simulateDelay({ data: mockStore.find((o) => o.id === Number(id)) });
  }
  const { data } = await httpClient.get(`/orders/${id}`);
  return data;
}

export function getMockOrdersSnapshot() {
  return mockStore;
}

export async function updateOrderStatus(id, status) {
  if (env.useMocks) {
    const numericId = Number(id);
    mockStore = mockStore.map((o) => (o.id === numericId ? { ...o, status } : o));
    persist();
    return simulateDelay({ data: mockStore.find((o) => o.id === numericId) });
  }
  const { data } = await httpClient.patch(`/orders/${id}/status`, { status });
  return data;
}
